File: nodejs/test/e2e/agent_and_compact_rpc.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { describe, expect, it } from "vitest";
import { approveAll } from "../../src/index.js";
import type { CustomAgentConfig } from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

describe("Agent Selection RPC", async () => {
    const { copilotClient: client } = await createSdkTestContext();

    it("should list available custom agents", async () => {
        const customAgents: CustomAgentConfig[] = [
            {
                name: "test-agent",
                displayName: "Test Agent",
                description: "A test agent",
                prompt: "You are a test agent.",
            },
            {
                name: "another-agent",
                displayName: "Another Agent",
                description: "Another test agent",
                prompt: "You are another agent.",
            },
        ];

        const session = await client.createSession({
            onPermissionRequest: approveAll,
            customAgents,
        });

        const result = await session.rpc.agent.list();
        expect(result.agents).toBeDefined();
        expect(Array.isArray(result.agents)).toBe(true);
        expect(result.agents.length).toBe(2);
        expect(result.agents[0].name).toBe("test-agent");
        expect(result.agents[0].displayName).toBe("Test Agent");
        expect(result.agents[0].description).toBe("A test agent");
        expect(result.agents[1].name).toBe("another-agent");

        await session.destroy();
    });

    it("should return null when no agent is selected", async () => {
        const customAgents: CustomAgentConfig[] = [
            {
                name: "test-agent",
                displayName: "Test Agent",
                description: "A test agent",
                prompt: "You are a test agent.",
            },
        ];

        const session = await client.createSession({
            onPermissionRequest: approveAll,
            customAgents,
        });

        const result = await session.rpc.agent.getCurrent();
        expect(result.agent).toBeNull();

        await session.destroy();
    });

    it("should select and get current agent", async () => {
        const customAgents: CustomAgentConfig[] = [
            {
                name: "test-agent",
                displayName: "Test Agent",
                description: "A test agent",
                prompt: "You are a test agent.",
            },
        ];

        const session = await client.createSession({
            onPermissionRequest: approveAll,
            customAgents,
        });

        // Select the agent
        const selectResult = await session.rpc.agent.select({ name: "test-agent" });
        expect(selectResult.agent).toBeDefined();
        expect(selectResult.agent.name).toBe("test-agent");
        expect(selectResult.agent.displayName).toBe("Test Agent");

        // Verify getCurrent returns the selected agent
        const currentResult = await session.rpc.agent.getCurrent();
        expect(currentResult.agent).not.toBeNull();
        expect(currentResult.agent!.name).toBe("test-agent");

        await session.destroy();
    });

    it("should deselect current agent", async () => {
        const customAgents: CustomAgentConfig[] = [
            {
                name: "test-agent",
                displayName: "Test Agent",
                description: "A test agent",
                prompt: "You are a test agent.",
            },
        ];

        const session = await client.createSession({
            onPermissionRequest: approveAll,
            customAgents,
        });

        // Select then deselect
        await session.rpc.agent.select({ name: "test-agent" });
        await session.rpc.agent.deselect();

        // Verify no agent is selected
        const currentResult = await session.rpc.agent.getCurrent();
        expect(currentResult.agent).toBeNull();

        await session.destroy();
    });

    it("should return empty list when no custom agents configured", async () => {
        const session = await client.createSession({ onPermissionRequest: approveAll });

        const result = await session.rpc.agent.list();
        expect(result.agents).toEqual([]);

        await session.destroy();
    });
});

describe("Session Compact RPC", async () => {
    const { copilotClient: client } = await createSdkTestContext();

    it("should compact session history after messages", async () => {
        const session = await client.createSession({ onPermissionRequest: approveAll });

        // Send a message to create some history
        await session.sendAndWait({ prompt: "What is 2+2?" });

        // Compact the session
        const result = await session.rpc.compaction.compact();
        expect(typeof result.success).toBe("boolean");
        expect(typeof result.tokensRemoved).toBe("number");
        expect(typeof result.messagesRemoved).toBe("number");

        await session.destroy();
    }, 60000);
});

File: go/internal/e2e/agent_and_compact_rpc_test.go
package e2e

import (
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
	"github.com/github/copilot-sdk/go/rpc"
)

func TestAgentSelectionRpc(t *testing.T) {
	cliPath := testharness.CLIPath()
	if cliPath == "" {
		t.Fatal("CLI not found. Run 'npm install' in the nodejs directory first.")
	}

	t.Run("should list available custom agents", func(t *testing.T) {
		client := copilot.NewClient(&copilot.ClientOptions{
			CLIPath:  cliPath,
			UseStdio: copilot.Bool(true),
		})
		t.Cleanup(func() { client.ForceStop() })

		if err := client.Start(t.Context()); err != nil {
			t.Fatalf("Failed to start client: %v", err)
		}

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			CustomAgents: []copilot.CustomAgentConfig{
				{
					Name:        "test-agent",
					DisplayName: "Test Agent",
					Description: "A test agent",
					Prompt:      "You are a test agent.",
				},
				{
					Name:        "another-agent",
					DisplayName: "Another Agent",
					Description: "Another test agent",
					Prompt:      "You are another agent.",
				},
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		result, err := session.RPC.Agent.List(t.Context())
		if err != nil {
			t.Fatalf("Failed to list agents: %v", err)
		}

		if len(result.Agents) != 2 {
			t.Fatalf("Expected 2 agents, got %d", len(result.Agents))
		}
		if result.Agents[0].Name != "test-agent" {
			t.Errorf("Expected first agent name 'test-agent', got %q", result.Agents[0].Name)
		}
		if result.Agents[0].DisplayName != "Test Agent" {
			t.Errorf("Expected first agent displayName 'Test Agent', got %q", result.Agents[0].DisplayName)
		}
		if result.Agents[1].Name != "another-agent" {
			t.Errorf("Expected second agent name 'another-agent', got %q", result.Agents[1].Name)
		}

		if err := client.Stop(); err != nil {
			t.Errorf("Expected no errors on stop, got %v", err)
		}
	})

	t.Run("should return null when no agent is selected", func(t *testing.T) {
		client := copilot.NewClient(&copilot.ClientOptions{
			CLIPath:  cliPath,
			UseStdio: copilot.Bool(true),
		})
		t.Cleanup(func() { client.ForceStop() })

		if err := client.Start(t.Context()); err != nil {
			t.Fatalf("Failed to start client: %v", err)
		}

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			CustomAgents: []copilot.CustomAgentConfig{
				{
					Name:        "test-agent",
					DisplayName: "Test Agent",
					Description: "A test agent",
					Prompt:      "You are a test agent.",
				},
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		result, err := session.RPC.Agent.GetCurrent(t.Context())
		if err != nil {
			t.Fatalf("Failed to get current agent: %v", err)
		}

		if result.Agent != nil {
			t.Errorf("Expected no agent selected, got %v", result.Agent)
		}

		if err := client.Stop(); err != nil {
			t.Errorf("Expected no errors on stop, got %v", err)
		}
	})

	t.Run("should select and get current agent", func(t *testing.T) {
		client := copilot.NewClient(&copilot.ClientOptions{
			CLIPath:  cliPath,
			UseStdio: copilot.Bool(true),
		})
		t.Cleanup(func() { client.ForceStop() })

		if err := client.Start(t.Context()); err != nil {
			t.Fatalf("Failed to start client: %v", err)
		}

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			CustomAgents: []copilot.CustomAgentConfig{
				{
					Name:        "test-agent",
					DisplayName: "Test Agent",
					Description: "A test agent",
					Prompt:      "You are a test agent.",
				},
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		// Select the agent
		selectResult, err := session.RPC.Agent.Select(t.Context(), &rpc.SessionAgentSelectParams{Name: "test-agent"})
		if err != nil {
			t.Fatalf("Failed to select agent: %v", err)
		}
		if selectResult.Agent.Name != "test-agent" {
			t.Errorf("Expected selected agent 'test-agent', got %q", selectResult.Agent.Name)
		}
		if selectResult.Agent.DisplayName != "Test Agent" {
			t.Errorf("Expected displayName 'Test Agent', got %q", selectResult.Agent.DisplayName)
		}

		// Verify getCurrent returns the selected agent
		currentResult, err := session.RPC.Agent.GetCurrent(t.Context())
		if err != nil {
			t.Fatalf("Failed to get current agent: %v", err)
		}
		if currentResult.Agent == nil {
			t.Fatal("Expected an agent to be selected")
		}
		if currentResult.Agent.Name != "test-agent" {
			t.Errorf("Expected current agent 'test-agent', got %q", currentResult.Agent.Name)
		}

		if err := client.Stop(); err != nil {
			t.Errorf("Expected no errors on stop, got %v", err)
		}
	})

	t.Run("should deselect current agent", func(t *testing.T) {
		client := copilot.NewClient(&copilot.ClientOptions{
			CLIPath:  cliPath,
			UseStdio: copilot.Bool(true),
		})
		t.Cleanup(func() { client.ForceStop() })

		if err := client.Start(t.Context()); err != nil {
			t.Fatalf("Failed to start client: %v", err)
		}

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			CustomAgents: []copilot.CustomAgentConfig{
				{
					Name:        "test-agent",
					DisplayName: "Test Agent",
					Description: "A test agent",
					Prompt:      "You are a test agent.",
				},
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		// Select then deselect
		_, err = session.RPC.Agent.Select(t.Context(), &rpc.SessionAgentSelectParams{Name: "test-agent"})
		if err != nil {
			t.Fatalf("Failed to select agent: %v", err)
		}

		_, err = session.RPC.Agent.Deselect(t.Context())
		if err != nil {
			t.Fatalf("Failed to deselect agent: %v", err)
		}

		// Verify no agent is selected
		currentResult, err := session.RPC.Agent.GetCurrent(t.Context())
		if err != nil {
			t.Fatalf("Failed to get current agent: %v", err)
		}
		if currentResult.Agent != nil {
			t.Errorf("Expected no agent selected after deselect, got %v", currentResult.Agent)
		}

		if err := client.Stop(); err != nil {
			t.Errorf("Expected no errors on stop, got %v", err)
		}
	})

	t.Run("should return empty list when no custom agents configured", func(t *testing.T) {
		client := copilot.NewClient(&copilot.ClientOptions{
			CLIPath:  cliPath,
			UseStdio: copilot.Bool(true),
		})
		t.Cleanup(func() { client.ForceStop() })

		if err := client.Start(t.Context()); err != nil {
			t.Fatalf("Failed to start client: %v", err)
		}

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		result, err := session.RPC.Agent.List(t.Context())
		if err != nil {
			t.Fatalf("Failed to list agents: %v", err)
		}

		if len(result.Agents) != 0 {
			t.Errorf("Expected empty agent list, got %d agents", len(result.Agents))
		}

		if err := client.Stop(); err != nil {
			t.Errorf("Expected no errors on stop, got %v", err)
		}
	})
}

func TestSessionCompactionRpc(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	if err := client.Start(t.Context()); err != nil {
		t.Fatalf("Failed to start client: %v", err)
	}

	t.Run("should compact session history after messages", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		// Send a message to create some history
		_, err = session.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "What is 2+2?",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		// Compact the session
		result, err := session.RPC.Compaction.Compact(t.Context())
		if err != nil {
			t.Fatalf("Failed to compact session: %v", err)
		}

		// Verify result has expected fields (just check it returned valid data)
		if result == nil {
			t.Fatal("Expected non-nil compact result")
		}
	})
}
